File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        
        Map<Integer,ArrayList<Integer>> map = new HashMap<>();
        for(int i=1; i<=N; i++) {
            map.put(i, new ArrayList<Integer>());
        }
        
        for(int i=0; i<M; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());

            ArrayList<Integer> list;
            
            if(map.containsKey(u)) list = map.get(u);
            else list = new ArrayList<Integer>();
            list.add(v);
            map.put(u, list);

            if(map.containsKey(v)) list = map.get(v);
            else list = new ArrayList<Integer>();
            list.add(u);
            map.put(v, list);
        }
        
        int answer = 0;
        boolean[] ch = new boolean[N+1];
        Queue<Integer> q = new LinkedList<>();
        
        for(int key : map.keySet()) {
            if(!ch[key]) {
                ch[key] = true;
                q.offer(key);
                while(!q.isEmpty()) {
                    int cur = q.poll();
                    ArrayList<Integer> list = map.get(cur);
                    for(int n : list) {
                        if(!ch[n]) {
                            q.offer(n);
                            ch[n] = true;
                        }
                    }
                }
                answer++;
            }
        }

        System.out.println(answer);
    }
}

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.swift
let nm = readLine()!.split(separator: " ").map{Int($0)!}
let N = nm[0]
let M = nm[1]

var dict:[Int:[Int]] = [:]
for n in 1...N { dict[n] = [] }

for _ in 0..<M {
    let uv = readLine()!.split(separator: " ").map{Int($0)!}
    let u = uv[0]
    let v = uv[1]
    
    if(dict[u] == nil) { dict[u] = [v] }
    else { dict[u]!.append(v) }
    
    if(dict[v] == nil) { dict[v] = [u] }
    else { dict[v]!.append(u) }
}

var answer = 0
var ch = Array(repeating: false, count: N+1)
var q:[Int] = []

for key in dict.keys {
    if(!ch[key]) {
        ch[key] = true
        q.append(key)
        
        while(!q.isEmpty) {
            let cur = q.removeFirst()
            for n in dict[cur] ?? [] {
                if(!ch[n]) {
                    ch[n] = true
                    q.append(n)
                }
            }
        }
        answer += 1
    }
}
print(answer)